### PR TITLE
feat: added mongoose support

### PR DESCRIPTION
**Pull request by OrrLumigo**
…performance

### CODE REVIEW

File: src/extender.js [hunk 1]
```diff
@@ -6,11 +6,11 @@ const noop = () => {};
 
 const isFunctionAlreadyWrapped = (fn) => fn && fn.__wrapped;
 
-export const hook = (module, funcName, options = {}, shimmerLib = shimmer) => {
+export const hook = (module, funcName, options = {}, initialContext = {}, shimmerLib = shimmer) => {
```
Comment (doriaviram) [inline src/extender.js:9]: Add test for that parameter (using that)

File: src/hooks/mongoose.js [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+  'findOneAndReplace',
+  'findOneAndRemove',
+];
+
+const ActiveConnectionDetails = (() => {
+  let _activeConnection = {};
+
+  const getActiveConnection = () => _activeConnection;
+
+  const updateFromConnectionString = (mssqlConnectionString) => {
```
Comment (doriaviram) [inline src/hooks/mongoose.js:35]: Don't duplicate all these parts...

File: src/hooks/mongoose.js [hunk 3]
```diff
@@ -0,0 +1,152 @@
+import { safeRequire } from '../utils/requireUtils';
+import * as logger from '../logger';
+import { hook, hookPromise } from '../extender';
+import { getRandomId, isPromise } from '../utils';
+import { SpansContainer, TracerGlobals } from '../globals';
+import { createMongoDbSpan, extendMongoDbSpan } from '../spans/mongoDbSpan';
+import { getCurrentTransactionId } from '../spans/awsSpan';
+
+const mongooseQueries = [
```
Comment (doriaviram) [inline src/hooks/mongoose.js:9]: BTW, Why our MongoDB instrumentation doesn't work on mongooses?

File: src/hooks/mongoose.js [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    {
+      started,
+    },
+    {
+      databaseName: connectionParameters.database,
+      commandName: extenderContext.funcName,
+      command: query,
+      mongoRequestId: 13,
+      mongoOperationId: undefined,
+      mongoConnectionId: 1,
```
Comment (doriaviram) [inline src/hooks/mongoose.js:111]: What's that magics numbers?

File: src/hooks/mongoose.js [hunk 5]
```diff
@@ -0,0 +1,152 @@
+import { safeRequire } from '../utils/requireUtils';
```
Comment (doriaviram) [inline src/hooks/mongoose.js:1]: new files is `.ts` only

**Review by doriaviram — CHANGES_REQUESTED** [review]
All these copy-pastes are smell like a design mistake.
Lets talk

File: src/hooks/mongoose.js (hunk 3 above)
Comment (OrrLumigo) [reply src/hooks/mongoose.js:9]: not sure why but it doesnt

File: src/hooks/mongoose.js (hunk 4 above)
Comment (OrrLumigo) [reply src/hooks/mongoose.js:111]: that need to be fixed.. 
but i dont get those fields in the hooks..

**Review by doriaviram — APPROVED** [review]

File: src/hooks/mongodb.js [hunk 6]
```diff
@@ -73,20 +73,26 @@ const onFailedHook = (event) => {
...
-  const mongoClient = mongoLib || safeRequire('mongodb');
-  if (mongoClient) {
-    logger.info('Starting to instrument MongoDB');
-    const listener = mongoClient.instrument({}, (err) => {
-      if (err) logger.warn('MongoDB instrumentation failed ', err);
+  const mongoClient = mongoLib ? mongoLib : safeRequire('mongodb');
+  const mongoosClients = safeRequire('node_modules/mongoose/node_modules/mongodb');
+  const mongoClients = [mongoClient, mongoosClients].filter(Boolean);
+  if (mongoClients.length > 0) {
+    mongoClients.forEach((mongoClient) => {
```
Comment (saartochner) [inline src/hooks/mongodb.js:80]: you don't need the `if` when you're doing `forEach` 🤔

File: src/hooks/mongodb.js [hunk 7]
```diff
@@ -73,20 +73,26 @@ const onFailedHook = (event) => {
...
+      if (mongoClient) {
```
Comment (saartochner) [inline src/hooks/mongodb.js:81]: you already filtered any falsy client

File: src/hooks/mongodb.js (hunk 6 above)
Comment (OrrLumigo) [reply src/hooks/mongodb.js:80]: this is for the log in the else

**Pull request merged**